crates/codegen/parser_runtime/src/templates/language.tera

```diff
@@ -0,0 +1,236 @@
+        {
+            let start = stream.position();
+            if self.{{ context.name | snake_case }}_next_token(stream) != Some(kind) {
+                stream.set_position(start);
+                return ParserResult::no_match(vec![kind]);
+            }
+            let end = stream.position();
+            return ParserResult::r#match(
+                vec![cst::Node::token(kind, stream.content(start.utf8..end.utf8))],
+                vec![],
+            );
+        }
+
+        #[allow(unused_assignments, unused_parens)]
+        pub fn {{ context.name | snake_case }}_next_token(&self, stream: &mut Stream) -> Option<TokenKind> {
+            let save = stream.position();
+            let mut furthest_position = stream.position();
+            let mut longest_token = TokenKind::SKIPPED; // just a marker
+
+            if let Some(kind) = {{ context.alpha_literal_scanner }} {
+                // Make sure that this is not the start of an identifier
+                if !self.identifier_part(stream) {
+                    furthest_position = stream.position();
+                    longest_token = kind;
+                }
+            }
+            stream.set_position(save);
+            
+            if let Some(kind) = {{ context.non_alpha_literal_scanner }} {
+                furthest_position = stream.position();
+                longest_token = kind;
+            }
+            stream.set_position(save);
+
+            macro_rules! longest_match {
+                ($( { $kind:ident = $function:ident } )*) => {
+                    $(
+                        if self.$function(stream) && stream.position() > furthest_position {
+                            furthest_position = stream.position();
+                            longest_token = TokenKind::$kind;
+                        }
+                        stream.set_position(save);
+                    )*
+                };
+            }
+
+            longest_match! {
+            {%- for name in context.compound_scanner_names -%}
+                {# Separator #}
+                { {{ name }} = {{ name | snake_case }} }
+            {%- endfor -%}
+                {# Separator #}
+            }
+
+            if longest_token != TokenKind::SKIPPED {
+                stream.set_position(furthest_position);
+                Some(longest_token)
+            } else {
+                None
+            }
+        }
+        {# Separator #}
+    {%- endfor -%}
+    {# Separator #}
+
+    /********************************************
+     *         Parser Functions
+     ********************************************/
+    {# Separator #}
+
+    {%- for function in code.parser_functions -%}
+        {# Separator #}
+        #[allow(unused_assignments, unused_parens)]
+        pub fn {{ function.0 | snake_case }}(&self, stream: &mut Stream) -> ParserResult { {{ function.1 }} }
```

Comment:
Just a note: this matches the Rust naming convention but I found it hurt greppability while I tried to navigate the codebase initially. I wonder if we can do something to retain the link between the generated functions and the grammar names, maybe using inlined wrappers? Just a thought, not sure if it's ultimately worth it.

Reply:
Do you mean keeping the PascalCase name for the function?

Reply:
I think we should ultimately put the EBNF for each rule above its method, to make it easier to match the implementation/structure. The EBNF uses PascalCase, so it should be greppable along other instances of the rule.
Happy to do this once I move the EBNF to the new structure. Thoughts?